lightning/src/ln/msgs.rs

```diff
@@ -960,33 +965,78 @@ impl Writeable for OnionHopData {
 	fn write<W: Writer>(&self, w: &mut W) -> Result<(), ::std::io::Error> {
 		w.size_hint(33);
 		match self.format {
-			OnionHopDataFormat::Legacy => 0u8.write(w)?,
+			OnionHopDataFormat::Legacy { short_channel_id } => {
+				0u8.write(w)?;
+				short_channel_id.write(w)?;
+				self.amt_to_forward.write(w)?;
+				self.outgoing_cltv_value.write(w)?;
+			},
+			OnionHopDataFormat::NonFinalNode { short_channel_id } => {
+				encode_varint_length_prefixed_tlv!(w, {
+					(2, self.amt_to_forward),
+					(4, self.outgoing_cltv_value),
+					(6, short_channel_id)
+				});
+			},
+			OnionHopDataFormat::FinalNode => {
+				encode_varint_length_prefixed_tlv!(w, {
+					(2, self.amt_to_forward),
+					(4, self.outgoing_cltv_value)
+				});
+			},
+		}
+		match self.format {
+			OnionHopDataFormat::Legacy { .. } => {
+				w.write_all(&[0;12])?;
+			},
+			_ => {},
 		}
-		self.short_channel_id.write(w)?;
-		self.amt_to_forward.write(w)?;
-		self.outgoing_cltv_value.write(w)?;
-		w.write_all(&[0;12])?;
 		Ok(())
 	}
 }
 
 impl<R: Read> Readable<R> for OnionHopData {
-	fn read(r: &mut R) -> Result<Self, DecodeError> {
-		Ok(OnionHopData {
-			format: {
-				let r: u8 = Readable::read(r)?;
-				if r != 0 {
-					return Err(DecodeError::UnknownVersion);
+	fn read(mut r: &mut R) -> Result<Self, DecodeError> {
+		use bitcoin::consensus::encode::{Decodable, Error, VarInt};
+		let v: VarInt = Decodable::consensus_decode(&mut r)
+			.map_err(|e| match e {
+				Error::Io(ioe) => DecodeError::from(ioe),
+				_ => DecodeError::InvalidValue
+			})?;
+		let (format, amt, cltv_value) = if v.0 != 0 {
+			let mut rd = FixedLengthReader { read: r, read_len: 0, max_len: v.0 };
+			let mut amt: u64 = 0;
+			let mut cltv_value: u32 = 0;
+			let mut short_id: Option<u64> = None;
+			decode_tlv!(&mut rd, {
+				(2, amt),
+				(4, cltv_value)
+			}, {
+				(6, short_id)
+			});
+			rd.eat_remaining().map_err(|_| DecodeError::ShortRead)?;
```

Comment:
I'm really not sure what you're suggesting here? Do you mean calling eat_remaining() in FixedLengthReader::drop() (which we can't do cause we have to sometimes return an Err)?

Reply:
For future types, yea, we probably should. Sadly, the "length" got overloaded so that 0 implies the legacy format, which is somewhat confusing, so I don't see an obvious+generic way to do this. Happy to change it if you see some better alternative.

Reply:
Couldn't the macros accept the length as a parameter?

Reply:
I mean I think in the future messages wont *have* a fixed length, they'll just contain a TLV and you will read to the end until you run out of TLV. This may be the only place that we use it, and, if not, I'm not sure what it'll look like. So I kinda agree with Arik.